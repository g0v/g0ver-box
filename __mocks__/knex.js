import knex from 'jest-mock-knex';

export { client } from 'jest-mock-knex';

export default knex;

process.setMaxListeners(0);
