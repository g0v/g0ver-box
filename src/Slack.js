import _ from 'lodash';
import slack from 'slack';

export const bot = slack.rtm.client();

const TOKEN = process.env.SLACK_TOKEN;

if (TOKEN) {
  bot.listen({ token: TOKEN });
}

bot.hello((message) => {
  console.log(`Got a message: ${message.type}`);
});

const call = (name, req = {}) => (
  new Promise((resolve, reject) => {
    const fn = _.get(slack, name);
    fn(
      { ...req, token: TOKEN },
      (err, res) => (err ? reject(err) : resolve(res))
    );
  })
);

export default {
  userInfo: async req => await call('users.info', req),
  postMessage: async req => await call('chat.postMessage', { ...req, as_user: true }),
  postMultiMessage: async (users, message) => _.map(users, user => bot.ws.send(JSON.stringify({
    id: `${user}::${Date.now()}`, type: 'message', channel: user, text: message,
  }))),
  channelInfo: async req => await call('channels.info', req),
  channelJoin: async req => await call('channels.join', req),
};
