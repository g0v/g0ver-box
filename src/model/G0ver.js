import _ from 'lodash';
import { ValueColumn, ListColumn } from 'graphql-tower';
import { Model } from './Database';
import GraphQLG0ver from '../type/GraphQLG0ver';

export default class G0ver extends Model {

  static tableName = 'g0ver';

  static columns = {
    username: new ValueColumn(),
    skills: new ListColumn(),
    slogan: new ValueColumn(),
  }

  static toKeyword({ skills }) {
    return _.toLower((skills || []).join(' '));
  }

  type = GraphQLG0ver;
}
