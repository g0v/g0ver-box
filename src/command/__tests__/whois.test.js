import { client } from 'knex';
import Slack from '../../Slack';
import G0ver from '../../model/G0ver';
import index from '../';

describe('whois command', () => {
  const data = { channel: 'D100', name: 'yutin', text: 'whois pm5' };

  it('when g0ver does not exist', async () => {
    client.mockReturnValueOnce([]);
    client.mockReturnValueOnce([]);
    await index(data);
    expect(client).toMatchSnapshot();
    expect(Slack.postMessage.mock.calls).toMatchSnapshot();
  });

  it('when g0ver is existed', async () => {
    G0ver.prime('pm5');
    client.mockReturnValueOnce([{ id: 'pm5', skills: ['react', 'harmonica'] }]);
    client.mockReturnValueOnce([]);
    await index(data);
    expect(client).toMatchSnapshot();
    expect(Slack.postMessage.mock.calls).toMatchSnapshot();
  });
});
