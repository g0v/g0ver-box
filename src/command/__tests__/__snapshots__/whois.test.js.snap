// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`whois command when g0ver does not exist 1`] = `
Array [
  "select * from g0ver where id in (pm5) and deleted_at is null",
  "select * from task where expired_at > DATE and deleted_at is null and user_id = pm5 limit 1",
]
`;

exports[`whois command when g0ver does not exist 2`] = `
Array [
  Array [
    Object {
      "channel": "D100",
      "text": "pm5
目前需要大家協助推坑
需要大使協助指導登錄技能",
    },
  ],
]
`;

exports[`whois command when g0ver is existed 1`] = `
Array [
  "select * from g0ver where id in (pm5) and deleted_at is null",
  "select * from task where expired_at > DATE and deleted_at is null and user_id = pm5 limit 1",
]
`;

exports[`whois command when g0ver is existed 2`] = `
Array [
  Array [
    Object {
      "channel": "D100",
      "text": "pm5
目前需要大家協助推坑
已登錄的技能 react, harmonica",
    },
  ],
]
`;
